lang/py/src/avro/io.py

```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+      None if (isinstance(datum, (int, long)) and INT_MIN_VALUE <= datum <= INT_MAX_VALUE) else \
+      BASE_ERR.format('int', type(datum))
+    )
   elif schema_type == 'long':
     if hasattr(expected_schema, 'logical_type'):
       if expected_schema.logical_type == constants.TIME_MICROS:
-        return isinstance(datum, datetime.time)
+        return ValidationResult(
+          None if isinstance(datum, datetime.time) else \
+          BASE_ERR.format('logicalType(Time-Micros), baseType(datetime.time)', type(datum))
+        )
       elif expected_schema.logical_type in [constants.TIMESTAMP_MILLIS, constants.TIMESTAMP_MICROS]:
-        return isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum)
-    return (isinstance(datum, (int, long))
-            and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE)
+        return ValidationResult(
+          None if isinstance(datum, datetime.datetime) and _is_timezone_aware_datetime(datum) else \
+          BASE_ERR.format(
+            'logicalType(Timestamp-Millis, Timestamp-Micros), baseType(datetime.datetime)',
+            type(datum)
+          )
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE) else \
+      BASE_ERR.format('long', type(datum))
+    )
   elif schema_type in ['float', 'double']:
-    return isinstance(datum, (int, long, float))
+    return ValidationResult(
+      None if isinstance(datum, (int, long, float)) else \
+      BASE_ERR.format('(float, double)', type(datum))
+    )
   # Check for int, float, long and decimal
   elif schema_type == 'fixed':
     if (hasattr(expected_schema, 'logical_type') and
         expected_schema.logical_type == constants.DECIMAL):
-      return isinstance(datum, Decimal)
-    return isinstance(datum, str) and len(datum) == expected_schema.size
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
+        BASE_ERR.format('logicalType(Decimal), baseType(fixed)', type(datum))
+      )
+    if not isinstance(datum, str):
+      return ValidationResult(BASE_ERR.format('str', type(datum)))
+    elif len(datum) != expected_schema.size:
+      return ValidationResult(
+        'Fixed type with invalid size, expected "{}" bytes found "{}" bytes'.format(
+          expected_schema.size, len(datum)
+        )
+      )
+    else:
+      return ValidationResult()
   elif schema_type == 'enum':
-    return datum in expected_schema.symbols
+    return ValidationResult(
+      None if datum in expected_schema.symbols else \
+      '"{}" is not among expected symbols {}'.format(datum, expected_schema.symbols)
+    )
   elif schema_type == 'array':
-    return (isinstance(datum, list) and
-      False not in [validate(expected_schema.items, d) for d in datum])
+    if not isinstance(datum, list):
+      return ValidationResult(BASE_ERR.format('list', type(datum)))
+    arr_items = [validate(expected_schema.items, d) for d in datum]
```

Comment:
nitpick: maybe call it `all_valid` or something similar so that it's easier to read in the following code: `if all_valid: return ValidationResult()`?